docs/_docs/plugins/generators.md

```diff
@@ -3,78 +3,112 @@ title: Generators
+
+Generators run after Jekyll has made an inventory of the existing content, and before the site is generated. Pages with
+front matter are stored as instances of [`Jekyll::Page`]({{ site.repository }}/blob/master/lib/jekyll/page.rb) and are
+available via `site.pages`. Static files become instances of
 [`Jekyll::StaticFile`]({{ site.repository }}/blob/master/lib/jekyll/static_file.rb)
-and are available via `site.static_files`. See
-[the Variables documentation page](/docs/variables/) and
-[`Jekyll::Site`]({{ site.repository }}/blob/master/lib/jekyll/site.rb)
-for details.
+and are available via `site.static_files`. See [the Variables documentation page](/docs/variables/) and
+[`Jekyll::Site`]({{ site.repository }}/blob/master/lib/jekyll/site.rb) for details.
 
-For instance, a generator can inject values computed at build time for template
-variables. In the following example, the template `reading.html` has two
-variables `ongoing` and `done` that are filled in the generator:
+In the following example, the generator will inject values computed at build time for template variables. The template
+named `reading.html` has two undefined variables `ongoing` and `done` that will be defined or assigned a value when
+the generator runs:
 
 ```ruby
 module Reading
   class Generator < Jekyll::Generator
     def generate(site)
-      ongoing, done = Book.all.partition(&:ongoing?)
+      book_data = site.data['books']
+      ongoing = book_data.select { |book| book['status'] == 'ongoing' }
+      done = book_data.select { |book| book['status'] == 'finished' }
+
+      # get template
+      reading = site.pages.find { |page| page.name == 'reading.html'}
 
-      reading = site.pages.detect {|page| page.name == 'reading.html'}
+      # inject data into template
       reading.data['ongoing'] = ongoing
       reading.data['done'] = done
     end
   end
 end
 ```
 
-The following example is a more complex generator that generates new pages. In this example, the generator will create a series of files under the `categories` directory for each category, listing the posts in each category using the `category_index.html` layout.
+The following example is a more complex generator that generates new pages.
+
+In this example, the aim of the generator is to create a page for each category registered in the `site`. The pages are
+created at runtime, so their contents, front matter and other attributes need to be designed by the plugin itself.
+* The pages are intended to render a list of all documents under a given category. So the basename of the rendered file
+would be better as `index.html`.
+* Having the ability to configure the pages via [front matter defaults](/docs/configuration/front-matter-defaults/)
+would be awesome! So assigning a particular `type` to these pages would be beneficial.
 
 ```ruby
-module Jekyll
-  class CategoryPageGenerator < Generator
+module SamplePlugin
```

Comment:
Is this still valid? I thought the plugin needs to be in the Jekyll module so that it gets run at build time.

Reply:
Yep. It *always has been valid*.
Any `*.rb` file inside `_plugins` will run at build time.